### ISSUE TITLE
Sizing New Types

### ISSUE THREAD

**Issue by nikivazou** [ucsd-progsys/liquidhaskell#825]
New Types are encoded as casts breaking termination checker. For example, 

```
newtype NT a = NT [a]

sizeNT :: NT a -> Int 
sizeNT (NT []) = 0 
sizeNT (NT (x:xs)) = 1 + sizeNT (NT xs)
```

returns with `No decreasing parameter`, while lifting `sizeNT` as a measure fails, as casts are not encoded into logic.

**Comment by gridaphobe** [comment]
Can we just use unsorted equality to encode the various `Cast`s in Core?

**Comment by nikivazou** [comment]
Not clear how. `NT` does not exists in the Core at all, it is just a cast.

**Comment by ranjitjhala** [comment]
So what core is generated for:

```haskell
newtype List a = LL [a]

cons :: a -> List a -> List a
cons x (LL xs) = LL (x:xs)
```

**Comment by gridaphobe** [comment]
```
cons :: forall a. a -> List a -> List a
[GblId, Arity=2, Caf=NoCafRefs, Str=DmdType]
cons =
  GHC.Types.:
  `cast` (forall (a :: <*>_N).
          <a>_R -> Sym Foo.N:List[0] <a>_N -> Sym Foo.N:List[0] <a>_N
          :: ((forall a. a -> [a] -> [a]) :: *)
             ~R#
             ((forall a. a -> List a -> List a) :: *))
```

It's actually really clever here, and just casts the `:` operator itself instead of casting the `List` to a `[]` and back.

**Comment by gridaphobe** [comment]
For reference, here's `sizeNT`

```
sizeNT [Occ=LoopBreaker] :: forall a. NT a -> Int
[GblId, Arity=1, Str=DmdType]
sizeNT =
  \ (@ a) (ds :: NT a) ->
    case ds `cast` (Foo.N:NT[0] <a>_N :: (NT a :: *) ~R# ([a] :: *))
    of _ [Occ=Dead] {
      [] -> GHC.Types.I# 0#;
      : x xs ->
        + @ Int
          GHC.Num.$fNumInt
          (GHC.Types.I# 1#)
          (sizeNT
             @ a
             (xs `cast` (Sym Foo.N:NT[0] <a>_N :: ([a] :: *) ~R# (NT a :: *))))
    }
```

@nikivazou why does it matter whether the `NT` constructor appears in the Core? A newtype should have the same size metric as its representation, we just need to explain that to LH.

**Comment by gridaphobe** [comment]
The more interesting problem (IMO) occurs when you want the newtype to have certain properties, e.g. a newtype for non-empty lists. Then we can't just ignore the `Cast`s, we have to realize that **certain** `Cast`s are the same as constructing or destructing data. This already happens to some extent with refined type classes, but the solution is a total hack (I just replace the `Cast` with an application of the constructor).

**Comment by nikivazou** [comment]
@gridaphobe I added invariants on new types (beta), see: https://github.com/ucsd-progsys/liquidhaskell/blob/b9a99c20c4164a4ccc8661cd509ef4ec344ed0ff/tests/pos/NewType.hs

I am not following your solution for termination check. Can you give me an example?

**Comment by gridaphobe** [comment]
Take your `sizeNT` function.

1. Let's assume the termination metric for `NT` is the same `len` we use for `[]`.
2. Encode `cast` with an unsorted equality.
3. Let's ANF the 1st `cast` to `let ys = ds `cast` ... in case ys of ...`
4. By 2, we know that `ys ~~ ds`
5. In the `: x xs` case we learn that `len ys = 1 + len xs`
6. Let's ANF the 2nd `cast` to `let zs = xs `cast` ... in sizeNT zs`
7. By 2, we know that `zs ~~ xs`
8. Now we need to prove that `len zs < len ds`. This is ill-sorted from fixpoint's view, but z3 won't care.

I guess the question is what to do with fixpoint here.

Perhaps for fixpoint's purposes we can unpack the newtypes and just give `NT`-typed values a `[]` sort? Then we could also use regular equality rather than unsorted.

**Comment by nikivazou** [comment]
I see, sorry my problem is more basic, I did not reach it for far. My problem is the following:

What is the (logical function) termination metric that is decreasing and how does the user encode it.

We cannot have a logical function of `NT`s because it does not exists in the core.

**Comment by ranjitjhala** [comment]
@nikivazou I don't follow what you mean

> We cannot have a logical function of NTs because it does not exists in
> the core.
>
It definitely exists in the core -- see the type signature for sizeNT or
the type ascribed to ds in the core pasted above?

Is the issue that there is no data constructor NT? (It having been replaced
by casts?)

**Comment by nikivazou** [comment]
Yes, the issue is that there is no data constructor, so it is not clear how to define a (logical) function  that measures the size of an `NT`. We cannot definitely define a measure.

**Comment by gridaphobe** [comment]
@nikivazou why should the size of `NT` be any different from the size of `[]`, they have the same structure?

Perhaps there are other measures that one might want to define differently for newtypes, but for the purpose of termination I don't see why we wouldn't just use the underlying structure.

**Comment by nikivazou** [comment]
I was going to say: because `sizeNT  :: xs:NT a -> Int \ [len xs] ` is not well formed. 

But, actually the above goes through our well formedness checker...

**Comment by nikivazou** [comment]
Since `sizeNT :: xs:NT a -> Int / [len xs]` is SAFE signature I am closing the issue. Still I am concerned because 
- `len xs` does not Haskell type check, but is well sorted in the logic
- is the first argument was `xs:[a]` then LH would automatically guess the correct termination metric, not with `xs:NT a` it does not guess any termination metric.

**Comment by gridaphobe** [comment]
It seems like a fluke that the explicit termination metric works, so let's keep this open.
